006-strimzi-ui.md

```diff
@@ -0,0 +1,187 @@
+# Add a UI to Strimzi for Kafka Administration
+
+This proposal is to add a Kafka administration UI to the Strimzi project.
+
+## Motivation
+
+Kafka out of the box provides a number of useful command tools to perform common administrative tasks such as creating topics, viewing consumer group status, and consuming record content. While these are helpful, they are not always the simplest or most convenient way of monitoring and performing administrative tasks on a Kafka cluster. In these cases a dedicated administration UI, focused on common administration tasks, would not only enable a user to complete their task more quickly and simply than previously possible, but also enable a more positive user experience of Kafka and its capabilities.
+
+## Proposal
+
+This proposal will describe a Kafka administration UI capability. It will describe example capabilities which could be offered, cover at a high level how the UI code itself could be implemented (with more low level detail to be provided as required), how this could be delivered, as well as possible future work and how this could integrate with Strimzi today.
+
+### UI implementation
+
+At a high level, I would propose a UI implemented as follows:
+
+- It would be a Javascript based UI (using [babel](https://babeljs.io/) to provide latest ECMAScript capabilities in a cross browser compatible manner), using the [React](https://reactjs.org/) framework
+- This UI is hosted and provided via an [Express](https://expressjs.com/) server
```

Comment:
Yes indeed! Build details and implementation (in more detail) can be found here: https://github.com/strimzi/strimzi-ui/pull/2 - which would have `webpack-dev-server` hosting, with a proxy for 'real' backend data

Reply:
First of all thank you @matthew-chirgwin for this initiative, I'm very keen to try this out. I just found the project strimzi-ui today and I might be missing a lot of context, so I would like to ask you a couple of things.

Is there a reason for going in this old school way with webpack, plain html and SASS, instead of Create React App for example? I also don't see why Express is necessary, wouldn't make more sense to deploy the static files served by Nginx?

For me as an user would be really nice if I can deploy the Kafka UI easily as a container in Kubernetes just writing a YAML using some CRDs (as the whole Strimzi project is designed to work with Kubernetes/Openshift), and the UI communicating directly to Kafka Bridge API. Since Kubernetes itself has a lot of security built-ins and I would deploy as a private container accessing it via port forward for example, and I wouldn't care too much about performance or security.

Reply:
Hi @neemiasjnr! Hopefully I can answer your questions :) 

In my experiece, Create React App (CRA) is awesome for getting something up and running quickly. It however (to enable ease of getting started) hides the configuration and tuning options you may need to change to suit a particular use case. In addition, it also brings a whole number of dependencies and boilerplate code that I do not think we will need to build a UI for Strimzi users.

To keep the UI minimal, and to enable the use cases we propose, I have cherry picked things like React, Webpack, Sass (used by CRA for instance), to build this UI. You can see the start of some of this now in [the Strimzi-ui repo](https://github.com/strimzi/strimzi-ui).

As per the proposal, we intend to use Express to serve the static content, but also own and mantain user session magement and handle client security. There are a number of useful 3rd party tools, such as Helmet, Redis and passport.js ready to use to enable this. It also allows for future expansion. For example, abstracting/combining multiple backend systems into one - enabling one UI managing multiple cluster use cases.

We are still working through the details, but the intent is to have exactly what you describe; a CR which references a Strimzi deployment/[backend](https://github.com/strimzi/proposals/pull/9) with some additional configuration options for items such as [security](https://github.com/strimzi/proposals/pull/13), and enabled/disabled capabilities. This is then reconciled by the cluster operator (for example - to be confirmed) and a deployment is created for you.

Hopefully this answers your questions, please let me know if there is anything more I can allborate on.